kernel: Allow k_poll on message queues

ngraves95 opened the pull request:
This commit adds the ability to use a message queue as a
k_poll object. It follows the same pattern as polling on
FIFOs.

This change has been proven in practice at Samsara.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/26728

Signed-off-by: Nick Graves <nicholas.graves@samsara.com>

andyross reviewed (approved):
This looks exactly correct to me, modulo my general horror with the way we've organized the type management in poll (that switch just shouldn't be needed, we should have a way to pend a thread on multiple wait queues and do all the type muckery just once at the outer level when iterating over the events).

But no reason not to merge this when it's clearly following the rules.  Just makes it more urgent to get k_poll cleaned up.

ngraves95 commented:
Sorry for what might be a stupid question, its my first time contributing to zephyr: I see that the pull request labeler build task is failing - do I need to rerun that or push a no-op change to make it rerun?

nashif reviewed (changes requested):
Need a test with coverage for all the introduced code before this can be merged.

ngraves95 commented:
> Need a test with coverage for all the introduced code before this can be merged.

I have added test. Can you please re-review and let me know if the tests are sufficient?

nashif commented:
> I have added test. Can you please re-review and let me know if the tests are sufficient?

thank you. Can you please put the test in its own commit?

Commit:
kernel: Allow k_poll on message queues

This commit adds the ability to use a message queue as a
k_poll object. It follows the same pattern as polling on
FIFOs.

This change has been proven in practice at Samsara.

Fixes: #26728

Signed-off-by: Nick Graves <nicholas.graves@samsara.com>

Commit:
tests/kernel: Add k_poll test for polling on message queues

Add tests for message queue polling functionality.

Fixes: #26728

Signed-off-by: Nick Graves <nicholas.graves@samsara.com>

ngraves95 commented:
> > I have added test. Can you please re-review and let me know if the tests are sufficient?
> 
> thank you. Can you please put the test in its own commit?

Done
